libvast/vast/system/source.hpp

```diff
@@ -307,9 +321,12 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
       //       implement an anycast downstream manager and use it for the
       //       source, because we mustn't duplicate data.
       auto& st = self->state;
-      if (st.sink) {
-        self->quit(make_error(ec::logic_error, "does not support multiple "
-                                               "sinks"));
+      if (st.sink.address() != sink.address()) {
+        self->quit(caf::make_error(ec::logic_error, "source does not support "
+                                                    "multiple sinks"));
+      } else if (st.sink) {
+        VAST_WARNING(self, "ignores request from", self->current_sender(),
+                     "to set same sink again");
```

Comment:
How can this happen? Seems like an invariant violation, so we should terminate at this point and not keep running.

Reply:
I've hit this once during testing and have never been able to reproduce it since. I'd like to leave it in until I can reason about it better.

Reply:
Can we make it an assertion or even hard abort? If it's such a weird error, we should make it more "in your face".

Reply:
I still want to see if I hit this again, which is why this for now logs the address of the sending actor. An assertion doesn't really allow for me to debug this here.

Reply:
I put some thought into this after a call with @dominiklohmann. I think an assertion is the right approach here, because in this case, the state of the source does not need to be changed and this alone is no reason to abort the whole process. We still want to see an error in debug and CI builds if it happens, and an assertions gives us exactly that.

Reply:
I'd argue it's reasonable to shut down the source, but just the source. The node itself should continue to live after the source shuts down.

Reply:
I'm suggesting this change:

```diff
diff --git a/libvast/vast/system/source.hpp b/libvast/vast/system/source.hpp
index f25fe5cc..b6110a9c 100644
--- a/libvast/vast/system/source.hpp
+++ b/libvast/vast/system/source.hpp
@@ -317,12 +317,9 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
       //       implement an anycast downstream manager and use it for the
       //       source, because we mustn't duplicate data.
       auto& st = self->state;
-      if (st.sink.address() != sink.address()) {
+      if (st.sink || st.sink.address() != sink.address()) {
         self->quit(caf::make_error(ec::logic_error, "source does not support "
                                                     "multiple sinks"));
-      } else if (st.sink) {
-        VAST_WARNING(self, "ignores request from", self->current_sender(),
-                     "to set same sink again");
         return;
       }
       st.sink = sink;
```